From 2371e5da683aa3a4d8cfc3fc31032440e16592d4 Mon Sep 17 00:00:00 2001
From: Andi Kleen <ak@linux.intel.com>
Date: Sat, 27 Jun 2020 20:50:06 -0700
Subject: Kbuild, lto: Generate debuginfo for temporary kallsyms links

gcc 10 changes the layout of the data segment when --strip-debug is
specified with LTO. This causes problems with kallsyms, which builds
temporary vmlinux with --strip-debug, and expects the final vmlinux
without --strip-debug to have the same symbol table.

To work around this problem generate debuginfo the temporary
links too when LTO is enabled. This will cause some extra IO.

Signed-off-by: Andi Kleen <ak@linux.intel.com>
---
 scripts/link-vmlinux.sh | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/scripts/link-vmlinux.sh b/scripts/link-vmlinux.sh
index 6eded325c8378..7961820c10a70 100755
--- a/scripts/link-vmlinux.sh
+++ b/scripts/link-vmlinux.sh
@@ -98,7 +98,9 @@ vmlinux_link()
 	shift
 
 	# The kallsyms linking does not need debug symbols included.
-	if [ "$output" != "${output#.tmp_vmlinux.kallsyms}" ] ; then
+	# except for LTO because gcc 10 LTO changes the layout of the data segment
+	# with --strip-debug
+	if [ "$output" != "${output#.tmp_vmlinux.kallsyms}" -a -z "$CONFIG_LTO" ] ; then
 		strip_debug=-Wl,--strip-debug
 	fi
 
-- 
cgit 1.2.3-1.el7

